qcoro/qcorotask.h

```diff
@@ -614,31 +650,55 @@ class Task {
     }
 
     template<typename ThenCallback, typename ErrorCallback, typename R = cb_invoke_result_t<ThenCallback, T>>
-    auto thenImpl(ThenCallback &&thenCallback, ErrorCallback &&errorCallback) -> std::conditional_t<detail::isTask_v<R>, R, Task<R>> {
+    auto thenImpl(QObject *context, ThenCallback &&thenCallback, ErrorCallback &&errorCallback) -> std::conditional_t<detail::isTask_v<R>, R, Task<R>> {
+        bool contextDeleted = false;
+        QMetaObject::Connection contextConn;
+        if (context) {
+            // Listen for the deletion of the context object.
+            // If the context object is deleted before the awaited task finishes,
+            // don't call the callback.
+            contextConn = QObject::connect(context, &QObject::destroyed, [&] {
+                contextDeleted = true;
+            });
+            // If we don't disconnect before returning,
+            // we create another problem like the one we are trying to fix.
+            // Then the lambda above would mutate contextDeleted in a deleted stack frame.
+        }
+
+        detail::Defer cleanup([contextConn]() {
+            QObject::disconnect(contextConn);
+        });
```

Comment:
Wouldn't it be simpler to simply create a `QPointer<QObject> contextPtr{context}` and then just check whether `contextPtr` is still valid before invoking the callback, instead of `contextDeleted`.

Reply:
Not sure, that would need adding a separate function for not passing a context.
Maybe that is a good idea anyway.

Reply:
You can create `const bool hasContext = (context != nullptr)` at the beginning and then check `!hasContext || !contextPtr.isNull()`. Probably still shorter and simpler than using the `destroyed` signal.